File: src/main/java/com/ascargon/rocketshow/util/ActionTrigger.java
package com.ascargon.rocketshow.util;

import com.ascargon.rocketshow.api.ActionHttp;
import com.ascargon.rocketshow.lighting.ActionLighting;
import com.ascargon.rocketshow.midi.ActionMidi;
import com.ascargon.rocketshow.raspberry.ActionRaspberryGpio;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.xml.bind.annotation.XmlElement;
import jakarta.xml.bind.annotation.XmlElementWrapper;
import jakarta.xml.bind.annotation.XmlElements;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
public abstract class ActionTrigger {

    private List<Action> actionList = new ArrayList<>();

    @XmlElementWrapper(name = "actionList")
    @XmlElements({@XmlElement(type = ActionNull.class, name = "actionNull"),
            @XmlElement(type = ActionSystem.class, name = "actionSystem"),
            @XmlElement(type = ActionTransport.class, name = "actionTransport"),
            @XmlElement(type = ActionMidi.class, name = "actionMidi"),
            @XmlElement(type = ActionLighting.class, name = "actionLighting"),
            @XmlElement(type = ActionRaspberryGpio.class, name = "actionRaspberryGpio"),
            @XmlElement(type = ActionHttp.class, name = "actionHttp")})
    @JsonProperty("actionList")
    public List<Action> getActionList() {
        return actionList;
    }

}

File: src/main/webapp/src/app/models/action-trigger.ts
import { Action } from "./action";
import { ActionHttp } from "./action-http";
import { ActionLighting } from "./action-lighting";
import { ActionMidi } from "./action-midi";
import { ActionNull } from "./action-null";
import { ActionRaspberryGpio } from "./action-raspberry-gpio";
import { ActionSystem } from "./action-system";
import { ActionTransport } from "./action-transport";

export class ActionTrigger {
  actionList: Action[] = [];

  constructor(data?: any) {
    if (!data) {
      return;
    }

    if (data.actionList) {
      for (let action of data.actionList) {
        let actionObject: Action;

        if (action.actionNull) {
          actionObject = new ActionNull(action.nullAction);
        } else if (action.actionSystem) {
          actionObject = new ActionSystem(action.actionSystem);
        } else if (action.actionTransport) {
          actionObject = new ActionTransport(action.actionTransport);
        } else if (action.actionMidi) {
          actionObject = new ActionMidi(action.actionMidi);
        } else if (action.actionLighting) {
          actionObject = new ActionLighting(action.actionLighting);
        } else if (action.actionRaspberryGpio) {
          actionObject = new ActionRaspberryGpio(action.actionRaspberryGpio);
        } else if (action.actionHttp) {
          actionObject = new ActionHttp(action.actionHttp);
        } else {
          console.warn("Uknown action type", action);
        }

        this.actionList.push(actionObject);
      }
    }
  }
}
